File: reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
    	if( head == NULL ) return head;

    	ListNode * pre = NULL;
    	ListNode * next = head->next;

    	while( true ){
    		head->next = pre;
    		if( next == NULL ) break;
    		pre = head;
    		head = next;
    		next = next->next;
    	}
    	return head;
    }
};

File: reverse-linked-list/reverse_linked_list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
	unless head.nil?
		_pre = nil
		_next = head.next

		while( true )
			head.next = _pre
			break if _next.nil?
			_pre = head
			head = _next
			_next = _next.next
		end
	end
	head
end